xhype/xhype/src/vmexit.rs

```diff
@@ -13,6 +16,104 @@ pub enum HandleResult {
     Next,
 }
 
+// Table 27-3
+#[inline]
+pub fn vmx_guest_reg(num: u64) -> X86Reg {
+    [
+        X86Reg::RAX,
+        X86Reg::RCX,
+        X86Reg::RDX,
+        X86Reg::RBX,
+        X86Reg::RSP,
+        X86Reg::RBP,
+        X86Reg::RSI,
+        X86Reg::RDI,
+        X86Reg::R8,
+        X86Reg::R9,
+        X86Reg::R10,
+        X86Reg::R11,
+        X86Reg::R12,
+        X86Reg::R13,
+        X86Reg::R14,
+        X86Reg::R15,
+    ][num as usize]
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// VMX_REASON_MOV_CR
+////////////////////////////////////////////////////////////////////////////////
+
+// Intel SDM Table 27-3.
+pub fn handle_cr(vcpu: &VCPU, _gth: &GuestThread) -> Result<HandleResult, Error> {
+    let qual = vcpu.read_vmcs(VMCS_RO_EXIT_QUALIFIC)?;
+    let ctrl_reg = match qual & 0xf {
+        0 => X86Reg::CR0,
+        3 => X86Reg::CR3,
+        4 => X86Reg::CR4,
+        8 => {
+            return Err(Error::Unhandled(
+                VMX_REASON_MOV_CR,
+                format!("access cr8: unimplemented"),
+            ))
+        }
+        _ => unreachable!(), // Table C-1. Basic Exit Reasons (Contd.)
+    };
+    let access_type = (qual >> 4) & 0b11;
+    let _lmsw_type = (qual >> 6) & 0b1;
+    let reg = vmx_guest_reg((qual >> 8) & 0xf);
+    let _source_data = (qual >> 16) & 0xffff;
+    match access_type {
+        0 => {
+            // move to cr
+            let mut new_value = vcpu.read_reg(reg)?;
+            match ctrl_reg {
+                X86Reg::CR0 => {
+                    new_value |= X86_CR0_NE;
+                    vcpu.write_vmcs(VMCS_CTRL_CR0_SHADOW, new_value)?;
+                }
+                _ => {
+                    return Err(Error::Unhandled(
+                        VMX_REASON_MOV_CR,
+                        format!("mov to {:?}: unimplemented", ctrl_reg),
+                    ));
+                }
+            }
+            vcpu.write_reg(ctrl_reg, new_value)?;
+
+            if ctrl_reg == X86Reg::CR0 || ctrl_reg == X86Reg::CR4 {
+                let cr0 = vcpu.read_reg(X86Reg::CR0)?;
+                let cr4 = vcpu.read_reg(X86Reg::CR4)?;
+                let mut efer = vcpu.read_vmcs(VMCS_GUEST_IA32_EFER)?;
+                let long_mode = cr0 & X86_CR0_PE > 0
```

Comment:
This reads a bit weird to me. These are one-bit fields. It may be just me but I'm more comfortable with code like
long_mode = cr0 & X86_CR0_PE != 0.

but this code is fine, I leave any change up to you, it is a matter of personal preference.

Reply:
I have changed `> 0` to `!= 0`.